sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StatefulAggregate.scala

```diff
@@ -180,11 +187,19 @@ case class StateStoreSaveExec(
 
           // Update and output modified rows from the StateStore.
           case Some(Update) =>
+
             new Iterator[InternalRow] {
-              private[this] val baseIterator = iter
+
+              // Filter late date using watermark if specified
+              private[this] val baseIterator = watermarkPredicate match {
```

Comment:
This is a very very good question. I had to go back and double check this with an improved test. But the behavior is correct in append mode even without the filter. This is because append mode and update mode are implemented differently. 

- *append mode*: consume the iterator and push all changes into state store, remove things that should be removed, and then create a new iterator from the removed elements. so late date (whose state have been dropped in the past) will get added and immediately removed before. The new iterator is smart enough to only filter those that got immediately added and removed. Note that this is blocking, the parent iterator is first consumed, and then a new iterator created.

- *update mode*: In contrast, update mode is not really blocking. A new iterator is immediately created that wraps the parent iterator. It one by one, consumes from parent iterator, updates state store, and returns the updated rows. So we need the filter out the rows with late data so that we dont emit anything for them.